src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java

```diff
@@ -4628,6 +4623,70 @@ public void execute()
       }
    }
    
+   @Override
+   public void onRequestDocumentSave(RequestDocumentSaveEvent event)
+   {
+      saveDocumentIds(event.getDocumentIds(), success -> 
+      {
+         if (SourceWindowManager.isMainSourceWindow())
+         {
+            server_.requestDocumentSaveCompleted(success,
+                  new VoidServerRequestCallback());
+         }
+      });
+   }
+   
+   @Override
+   public void onRequestDocumentClose(RequestDocumentCloseEvent event)
+   {
+      JsArrayString ids = event.getDocumentIds();
+      Command closeEditors = () ->
+      {
+         // Close each of the requested tabs
+         for (EditingTarget target: editors_)
+         {
+           if (JsArrayUtil.jsArrayStringContains(ids, target.getId()))
```

Comment:
That's a good point; I don't think it will. I'll add some safeguards (and a short circuit; there's no point in doing the whole waitformethod to close zero documents!)